crates/runtime/src/instance.rs

```diff
@@ -703,6 +756,55 @@ impl Instance {
         // dropping a non-passive segment is a no-op (not a trap).
     }
 
+    /// Get a table by index regardless of whether it is locally-defined
+    /// or an imported, foreign table. Ensure that the given range of
+    /// elements in the table is lazily initialized.  We define this
+    /// operation all-in-one for safety, to ensure the lazy-init
+    /// happens.
+    ///
+    /// Takes an `Iterator` for the index-range to lazy-initialize,
+    /// for flexibility. This can be a range, single item, or empty
+    /// sequence, for example. The iterator should return indices in
+    /// increasing order, so that the break-at-out-of-bounds behavior
+    /// works correctly.
+    pub(crate) fn get_table_with_lazy_init(
+        &mut self,
+        table_index: TableIndex,
+        range: impl Iterator<Item = u32>,
+    ) -> *mut Table {
+        let (idx, instance) = self.get_defined_table_index_and_instance(table_index);
+        let elt_ty = instance.tables[idx].element_type();
+
+        if elt_ty == TableElementType::Func {
+            let len = instance.tables[idx].size();
+            for i in range.take_while(|i| *i < len) {
+                let value = instance.tables[idx].get(i);
```

Comment:
This seems to use a fallible `get` in combination with `take_while` which ends up doing two rounds of bounds-checks? Could this instead always do `get` and `break` on out-of-bounds?

Reply:
Ah, yes, better!